apps/dashboard/src/components/workflow-editor/steps/email/html-editor.tsx

```diff
@@ -19,6 +23,85 @@ export function HtmlEditor({ value, onChange }: HtmlEditorProps) {
   const { step, digestStepBeforeCurrent } = useWorkflow();
   const { variables, isAllowedVariable } = useParseVariables(step?.variables, digestStepBeforeCurrent?.stepId);
 
+  const enhancedLiquidCompletionSource = useCallback((context: CompletionContext) => {
+    const result = liquidCompletionSource()(context);
+    if (!result) return null;
+
+    return {
+      ...result,
+      options: result?.options.map(
+        (option) =>
+          ({
+            ...option,
+            apply: (view: EditorView, completion: CompletionOption, from: number, to: number) => {
+              // Only apply to property completions, for example {{ forloop.first }}, where first is a property of forloop
+              if (completion.type !== 'property') {
+                return;
+              }
+
+              const selectedValue = completion.label;
+
+              const content = view.state.doc.toString();
+              const afterCursor = content.slice(to);
+
+              // Ensure proper {{ }} wrapping
+              const needsClosing = !afterCursor.startsWith('}}');
+
+              const wrappedValue = `${selectedValue}${needsClosing ? '}}' : ''}`;
+
+              // Calculate the final cursor position
+              // Add 2 if we need to account for closing brackets
+              const finalCursorPos = from + wrappedValue.length + (needsClosing ? 0 : 2);
+
+              view.dispatch({
+                changes: { from, to, insert: wrappedValue },
+                selection: { anchor: finalCursorPos },
+              });
+
+              return true;
+            },
+          }) as Completion
+      ),
+    };
+  }, []);
```

Comment:
This functionality autocompletes the liquid properties, for example, when the user typed `{{forloop.fi}}` and then picked the `first` property from the suggestions, the whole `{{forloop.first}}` will be inserted and converted to pill.

Reply:
In the future it will be moved to the `Editor` when we decide that we enable support for Liquid on all control values